front-end-tools/src/reading_from_blockchain.ts

```diff
@@ -0,0 +1,146 @@
+import {
+    toBuffer,
+    ConcordiumGRPCClient,
+    deserializeReceiveReturnValue,
+    serializeUpdateContractParameters,
+    ModuleReference,
+} from '@concordium/web-sdk';
+
+import { CONTRACT_SUB_INDEX } from './constants';
+
+export async function getContractInfo(rpcClient: ConcordiumGRPCClient | undefined, contractIndex: bigint) {
+    if (rpcClient === undefined) {
+        throw new Error(`rpcClient undefined`);
+    }
+    if (contractIndex === undefined) {
+        throw new Error(`Set smart contract index`);
+    }
+
+    const info = await rpcClient.getInstanceInfo({ index: contractIndex, subindex: CONTRACT_SUB_INDEX });
+
+    // Removing the `init_` prefix.
+    const contractName = info.name.substring(5);
+
+    // Removing the `contractName.` prefix.
+    const methods = info.methods.map((element) => element.substring(contractName.length + 1));
+
+    const returnValue = { contractName, methods, sourceModule: info.sourceModule };
+    return returnValue;
+}
+
+export async function getEmbeddedSchema(
+    rpcClient: ConcordiumGRPCClient | undefined,
+    moduleRef: ModuleReference | undefined
+) {
+    if (rpcClient === undefined) {
+        throw new Error(`rpcClient undefined`);
+    }
+    if (moduleRef === undefined) {
+        throw new Error(`Set module ref`);
+    }
+
+    return rpcClient.getEmbeddedSchema(moduleRef);
+}
+
+export async function read(
+    rpcClient: ConcordiumGRPCClient | undefined,
+    contractName: string | undefined,
```

Comment:
Why does this need to take parameters that are `| undefined`?

Reply:
Removed the `| undefined` at `contractName`.

Many other parameters return type errors if you remove the `|undefined`. E.g. the `entrypoint` parameter requires it because of this line:
```
form.setValue('entryPointName', e?.value);
```